v2/navigation.jsx

```diff
@@ -221,37 +223,34 @@ export function Navigation(props) {
 
     // reflect to the URL
     useEffect(() => {
-        // only act on navigations when settled
-        if (state.state !== 'settled') {
-            return
-        }
+        /**
+         * 'popstateHandler' is invoked on back AND forward navigations
+         * - To detect if it's a 'forward' intention, we look at the current 'stack' in the url params
+         *   and compare it to our local state. So if the url stack has 1 more item than our state, we know
+         *   it's a 'forward' action.
+         * - otherwise, it's just a 'back' action, so we can `pop` an item from the stack as usual
+         */
+        function popstateHandler() {
+            const currentUrlParams = new URLSearchParams(location.href)
+            const currentURLStack = currentUrlParams.getAll('stack')
+            const navigationIntentionIsForwards = currentURLStack.length > state.stack.length
 
-        if (state.via === 'push') {
-            const url = new URL(window.location.href)
-            url.searchParams.delete('stack')
-            for (let string of state.stack) {
-                url.searchParams.append('stack', string)
-            }
-            for (let [key, value] of Object.entries(state.params)) {
-                url.searchParams.set(key, value)
+            if (navigationIntentionIsForwards) {
+                const lastEntry = currentURLStack[currentURLStack.length - 1]
+                if (isScreenName(lastEntry)) {
+                    dispatch({ type: 'push', name: lastEntry, opts: { animate: props.animate && isAndroid() } })
+                }
+            } else {
+                dispatch({ type: 'pop', opts: { animate: props.animate && isAndroid() } })
             }
-            window.history.pushState({}, '', url)
         }
 
-        if (state.via === 'pop') {
-            window.history.go(-1)
-        }
-
-        function handler() {
-            dispatch({ type: 'pop', opts: { animate: props.animate } })
-        }
-
-        window.addEventListener('popstate', handler)
+        window.addEventListener('popstate', popstateHandler)
```

Comment:
This now handles forwards *and* back navigations - the dashboard has never supported the 'forward' swipe, but since I was here changing the navigation, it makes sense to complete it.

Reply:
It's interesting that forward navigations end up here too. I thought `history.pushState()` don't trigger popstate.. what am I missing?

Reply:
This catches the 'forward' button in the browser's chrome (and forward swipes in iOS).